File: scripts/registry/generate-service-index.mjs
/**
 * Regenerate docs/registry/services-index.md from services-registry.yaml
 *
 * Usage (from repo root):
 *   cd scripts/registry && npm install && npm run generate
 */

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import yaml from "js-yaml";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, "../..");
const REGISTRY_PATH = path.join(ROOT, "docs/registry/services-registry.yaml");
const OUTPUT_PATH = path.join(ROOT, "docs/registry/services-index.md");

function main() {
  if (!fs.existsSync(REGISTRY_PATH)) {
    console.error("Missing:", REGISTRY_PATH);
    process.exit(1);
  }
  const data = yaml.load(fs.readFileSync(REGISTRY_PATH, "utf8"));
  const version = data.registry_version ?? "?";
  const libs = [...(data.libraries ?? [])].sort((a, b) =>
    (a.maven_module ?? "").localeCompare(b.maven_module ?? "")
  );
  const services = [...(data.services ?? [])].sort((a, b) =>
    (a.maven_module ?? "").localeCompare(b.maven_module ?? "")
  );

  const generated = "deterministic-registry-generation";

  const lines = [
    "# Impilo service index (generated)",
    "",
    `**Registry version:** \`${version}\`  `,
    `**Generated:** ${generated}  `,
    "**Source:** [`services-registry.yaml`](./services-registry.yaml)  ",
    "",
    "Regenerate:",
    "",
    "```bash",
    "cd scripts/registry && npm install && npm run generate",
    "```",
    "",
    "---",
    "",
    "## Libraries (Maven reactor, not deployable)",
    "",
    "| ID | Maven module | Path |",
    "|----|--------------|------|",
  ];

  for (const e of libs) {
    const mid = e.maven_module ?? "";
    const p =
      e.path ?? (mid.startsWith("libs/") ? mid : `services/${mid}`);
    lines.push(`| \`${e.id ?? ""}\` | \`${mid}\` | \`${p}\` |`);
  }

  lines.push(
    "",
    "---",
    "",
    "## Services & agents",
    "",
    "| ID | Maven module | Primary plane | Domain | Secondary planes | Sovereign group | Product names | Protocol | Port | SoR responsibilities | Forbidden responsibilities | Consumes from | Exposes to | Impl status | Frontend wiring | API contract | Authz/Audit | Observability | Production status |",
    "|----|--------------|---------------|--------|------------------|-----------------|---------------|----------|------|----------------------|----------------------------|---------------|------------|-------------|-----------------|-------------|-------------|---------------|-------------------|"
  );

  for (const e of services) {
    const names = (e.product_names ?? []).length ? e.product_names.join(", ") : "—";
    const sg = e.sovereign_group ?? (e.sovereign ? "yes" : "—");
    const port = e.default_http_port != null ? String(e.default_http_port) : "—";
    const sec = (e.secondary_planes ?? []).length ? e.secondary_planes.join(", ") : "—";
    const sor = (e.system_of_record_for ?? []).length ? e.system_of_record_for.join("; ") : "—";
    const forb = (e.forbidden_responsibilities ?? []).length ? e.forbidden_responsibilities.join("; ") : "—";
    const upstream = (e.consumes_from ?? []).length ? e.consumes_from.join(", ") : "—";
    const downstream = (e.exposes_to ?? []).length ? e.exposes_to.join(", ") : "—";
    lines.push(
      `| \`${e.id ?? ""}\` | \`${e.maven_module ?? ""}\` | ${e.primary_plane ?? e.plane ?? ""} | ${e.domain ?? "—"} | ${sec} | ${sg} | ${names} | ${e.primary_protocol ?? ""} | ${port} | ${sor} | ${forb} | ${upstream} | ${downstream} | ${e.implementation_status ?? "—"} | ${e.frontend_wiring_status ?? "—"} | ${e.api_contract_status ?? "—"} | ${e.authz_audit_status ?? "—"} | ${e.observability_status ?? "—"} | ${e.production_status ?? "—"} |`
    );
  }

  lines.push(
    "",
    "---",
    "",
    "*Sovereign* marks membership in the nine national dual-mode boundaries (see `TODO.md` / doctrine). " +
      "*Group* is the logical sovereign name when one product spans multiple modules.",
    ""
  );

  fs.mkdirSync(path.dirname(OUTPUT_PATH), { recursive: true });
  fs.writeFileSync(OUTPUT_PATH, lines.join("\n"), "utf8");
  console.log("Wrote", path.relative(ROOT, OUTPUT_PATH));
}

main();

File: scripts/registry/generate-service-index.py
#!/usr/bin/env python3
"""
Regenerate docs/registry/services-index.md from docs/registry/services-registry.yaml.

Preferred (Node, same output as CI/local docs):
  cd scripts/registry && npm install && npm run generate

Optional (Python):
  pip install -r scripts/registry/requirements.txt
  python scripts/registry/generate-service-index.py
"""

from __future__ import annotations

import sys
from datetime import datetime, timezone
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
REGISTRY_PATH = ROOT / "docs" / "registry" / "services-registry.yaml"
OUTPUT_PATH = ROOT / "docs" / "registry" / "services-index.md"


def main() -> int:
    try:
        import yaml  # type: ignore
    except ImportError:
        print("Install PyYAML: pip install -r scripts/registry/requirements.txt", file=sys.stderr)
        return 1

    if not REGISTRY_PATH.is_file():
        print(f"Missing registry: {REGISTRY_PATH}", file=sys.stderr)
        return 1

    data = yaml.safe_load(REGISTRY_PATH.read_text(encoding="utf-8"))
    version = data.get("registry_version", "?")
    libs = data.get("libraries") or []
    services = data.get("services") or []

    def sort_key(entry: dict) -> str:
        return entry.get("maven_module") or entry.get("id") or ""

    libs = sorted(libs, key=sort_key)
    services = sorted(services, key=sort_key)

    lines: list[str] = [
        "# Impilo service index (generated)",
        "",
        f"**Registry version:** `{version}`  ",
        f"**Generated:** {datetime.now(timezone.utc).strftime('%Y-%m-%d %H:%M UTC')}  ",
        "**Source:** [`services-registry.yaml`](./services-registry.yaml)  ",
        "",
        "Regenerate:",
        "",
        "```bash",
        "pip install -r scripts/registry/requirements.txt",
        "python scripts/registry/generate-service-index.py",
        "```",
        "",
        "---",
        "",
        "## Libraries (Maven reactor, not deployable)",
        "",
        "| ID | Maven module | Path |",
        "|----|--------------|------|",
    ]
    for e in libs:
        mid = e.get("maven_module", "")
        path = e.get("path", f"services/{mid}" if not mid.startswith("libs/") else mid)
        lines.append(f"| `{e.get('id', '')}` | `{mid}` | `{path}` |")
    lines.extend(
        [
            "",
            "---",
            "",
            "## Services & agents",
            "",
            "| ID | Maven module | Primary plane | Domain | Secondary planes | Sovereign group | Product names | Protocol | Port | SoR responsibilities | Forbidden responsibilities | Consumes from | Exposes to | Impl status | Frontend wiring | API contract | Authz/Audit | Observability | Production status |",
            "|----|--------------|---------------|--------|------------------|-----------------|---------------|----------|------|----------------------|----------------------------|---------------|------------|-------------|-----------------|-------------|-------------|---------------|-------------------|",
        ]
    )
    for e in services:
        names = e.get("product_names") or []
        names_s = ", ".join(names) if names else "—"
        sg = e.get("sovereign_group") or ("yes" if e.get("sovereign") else "—")
        port = e.get("default_http_port")
        port_s = str(port) if port is not None else "—"
        sec_s = ", ".join(e.get("secondary_planes") or []) or "—"
        sor_s = "; ".join(e.get("system_of_record_for") or []) or "—"
        forb_s = "; ".join(e.get("forbidden_responsibilities") or []) or "—"
        upstream_s = ", ".join(e.get("consumes_from") or []) or "—"
        downstream_s = ", ".join(e.get("exposes_to") or []) or "—"
        lines.append(
            f"| `{e.get('id', '')}` | `{e.get('maven_module', '')}` | {e.get('primary_plane', e.get('plane', ''))} | {e.get('domain', '—')} | {sec_s} | {sg} | "
            f"{names_s} | {e.get('primary_protocol', '')} | {port_s} | {sor_s} | {forb_s} | {upstream_s} | {downstream_s} | "
            f"{e.get('implementation_status', '—')} | {e.get('frontend_wiring_status', '—')} | {e.get('api_contract_status', '—')} | "
            f"{e.get('authz_audit_status', '—')} | {e.get('observability_status', '—')} | {e.get('production_status', '—')} |"
        )
    lines.append("")
    lines.append("---")
    lines.append("")
    lines.append("*Sovereign* marks membership in the nine national dual-mode boundaries (see `TODO.md` / doctrine). "
                 "*Group* is the logical sovereign name when one product spans multiple modules.")
    lines.append("")

    OUTPUT_PATH.parent.mkdir(parents=True, exist_ok=True)
    OUTPUT_PATH.write_text("\n".join(lines), encoding="utf-8")
    print(f"Wrote {OUTPUT_PATH.relative_to(ROOT)}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
